processor/src_hydration_stage.go

```diff
@@ -0,0 +1,192 @@
+							obskit.SourceID(string(sourceId)),
+							obskit.Error(marshallErr),
+						)
+						return marshallErr
+					}
+					if marshalledPayload != nil {
+						originalESJob := eventSchemaJobsByJobID[jobId]
+						message.eventSchemaJobsBySourceId[sourceId] = append(message.eventSchemaJobsBySourceId[sourceId], &jobsdb.JobT{
+							UUID:         originalESJob.UUID,
+							UserID:       originalESJob.UserID,
+							Parameters:   originalESJob.Parameters,
+							CustomVal:    originalESJob.CustomVal,
+							EventPayload: marshalledPayload,
+							CreatedAt:    originalESJob.CreatedAt,
+							ExpireAt:     originalESJob.ExpireAt,
+							WorkspaceId:  originalESJob.WorkspaceId,
+						})
+					}
+				}
+			}
+			return nil
+		})
+	}
+
+	// Wait for all goroutines to complete and check for errors
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+
+	return &preTransformationMessage{
+		partition:                     message.partition,
+		subJobs:                       message.subJobs,
+		archivalJobs:                  message.archivalJobs,
+		eventSchemaJobsBySourceId:     message.eventSchemaJobsBySourceId,
+		connectionDetailsMap:          message.connectionDetailsMap,
+		statusDetailsMap:              message.statusDetailsMap,
+		enricherStatusDetailsMap:      message.enricherStatusDetailsMap,
+		botManagementStatusDetailsMap: message.botManagementStatusDetailsMap,
+		eventBlockingStatusDetailsMap: message.eventBlockingStatusDetailsMap,
+		destFilterStatusDetailMap:     message.destFilterStatusDetailMap,
+		reportMetrics:                 message.reportMetrics,
+		totalEvents:                   message.totalEvents,
+		groupedEventsBySourceId:       message.groupedEventsBySourceId,
+		eventsByMessageID:             message.eventsByMessageID,
+		jobIDToSpecificDestMapOnly:    message.jobIDToSpecificDestMapOnly,
+		statusList:                    message.statusList,
+		jobList:                       message.jobList,
+		sourceDupStats:                message.sourceDupStats,
+		dedupKeys:                     message.dedupKeys,
+	}, nil
+}
+
+func (proc *Handle) hydrate(ctx context.Context, source *backendconfig.SourceT, events []types.TransformerEvent) ([]types.TransformerEvent, error) {
+	req := types.SrcHydrationRequest{
+		Source: types.SrcHydrationSource{
+			ID:               source.ID,
+			Config:           source.Config,
+			WorkspaceID:      source.WorkspaceID,
+			SourceDefinition: source.SourceDefinition,
+			InternalSecret:   source.InternalSecret,
+		},
+	}
+	req.Batch = lo.Map(events, func(event types.TransformerEvent, _ int) types.SrcHydrationEvent {
+		return types.SrcHydrationEvent{
+			ID:    strconv.FormatInt(event.Metadata.JobID, 10),
+			Event: event.Message,
+		}
+	})
+	resp, err := proc.transformerClients.SrcHydration().Hydrate(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	jobIDToEventMap := lo.SliceToMap(events, func(event types.TransformerEvent) (string, types.TransformerEvent) {
```

Comment:
we are already iterating over the events for preparing the batch, can't we build that map then as well and skip yet another iteration?